### PR TITLE
fix(ci): catch multi-line and callback setTimeout patterns in check-test-timeouts (fixes #1632, fixes #1633)

### LINKED ISSUE #1632: fix(ci): check-test-timeouts regex misses setTimeout(() => callback(), N) pattern
## Background

#1530 added `scripts/check-test-timeouts.ts` to ban fixed-delay `setTimeout` in `*.spec.ts` files. Discovered during QA of PR #1628.

## Problem

The detection regex `/\bsetTimeout\s*\([^)]*[0-9]+[^)]*\)/` uses `[^)]*` which stops matching at the **first `)` inside the call**. Any `setTimeout` whose first argument is an arrow function is silently skipped:

```
FLAGGED  setTimeout(r, 50)                              ← caught
missed   setTimeout(() => r(null), 50)                 ← NOT caught
missed   setTimeout(() => {}, 60_000)                  ← NOT caught
```

`packages/daemon/src/claude-session/ws-server.spec.ts:1016` and `:3305` contain `new Promise<null>((r) => setTimeout(() => r(null), 50))` — genuine fixed-delay waits that bypass the new lint check.

## Proposed Fix

Options:
1. Change `[^)]*` to `[\s\S]*?` and limit the full match to a single logical line (requires multi-pass or a smarter extractor)
2. Parse the argument list properly (extract the last numeric arg) instead of relying on a single regex
3. Use AST-based analysis (e.g. `bun-ast` or `acorn`) to reliably detect the second argument

Also fix `ws-server.spec.ts:1016,:3305` to use `Bun.sleep(50)` or a proper poll-with-timeout, and add a `shouldMatch` test case for `setTimeout(() => doThing(), 50)` so the gap is covered going forward.

## Files
- `scripts/check-test-timeouts.ts:33` — VIOLATION_PATTERN
- `scripts/check-test-timeouts.spec.ts:17` — shouldMatch array missing arrow-fn form
- `packages/daemon/src/claude-session/ws-server.spec.ts:1016,3305` — undetected violations

### LINKED ISSUE #1633: fix(ci): check-test-timeouts misses multi-line setTimeout calls
## Background

#1530 added `scripts/check-test-timeouts.ts`. Discovered during QA of PR #1628 (Copilot comment #3133804220).

## Problem

The scan in `scripts/check-test-timeouts.ts:57` splits file content on `\n` and matches line-by-line. A `setTimeout` call split across multiple lines (common after auto-formatting) is invisible to the check:

```typescript
// This would NOT be flagged:
await new Promise((r) =>
  setTimeout(r, 50)
);
```

## Proposed Fix

Read the file as a whole string and use a regex with the `m`/`s` flag, or use a proper AST-based approach. At minimum, join continuation lines before testing the pattern.

## Files
- `scripts/check-test-timeouts.ts` — `scanDir` function, line-by-line matching

### PR DESCRIPTION
**Pull request by theshadow27**
## Summary

- Adds `findViolations(content: string)` that scans a full file string with paren-depth tracking across newlines, replacing the previous line-by-line approach. This catches `setTimeout` calls whose arguments are split across multiple lines (closes #1633) as well as the already-handled arrow-function callback form (closes #1632, which was partly addressed in #1628 but the coverage gap was documented).
- `hasFixedDelay` is kept and exported — existing per-line unit tests continue to use it unchanged.
- `scanDir` now delegates to `findViolations` instead of the old line-splitting loop.
- 6 new spec cases added for `findViolations`: multi-line delay argument, arrow-fn callback on next line, variable delay (no flag), single-arg (no flag), correct line numbers for multiple violations.

## Test plan

- [x] `bun test scripts/check-test-timeouts.spec.ts` — 25 pass (19 original + 6 new)
- [x] `bun scripts/check-test-timeouts.ts` — no violations found in codebase
- [x] `bun typecheck` — clean
- [x] `bun lint` — clean

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: scripts/check-test-timeouts.ts [hunk 1]
```diff
@@ -86,6 +86,53 @@ interface Violation {
...
+
+    const args = content.slice(parenOpen + 1, i - 1);
+
+    let argDepth = 0;
+    let lastComma = -1;
+    for (let j = 0; j < args.length; j++) {
+      const c = args[j];
+      if (c === "(" || c === "[" || c === "{") argDepth++;
+      else if (c === ")" || c === "]" || c === "}") argDepth--;
+      else if (c === "," && argDepth === 0) lastComma = j;
```
[the remark this answers is not in the document]
Comment (theshadow27) [reply scripts/check-test-timeouts.ts]: Pre-existing gap — the old line-by-line regex also missed the 3-arg form. Not a regression introduced by this PR; the 3-arg setTimeout is rare in test code. Filing as a separate issue.

File: scripts/check-test-timeouts.spec.ts [hunk 2]
```diff
@@ -44,3 +44,62 @@ describe("check-test-timeouts hasFixedDelay", () => {
     });
   }
 });
+
+describe("check-test-timeouts findViolations (multi-line)", () => {
+  test("catches single-line setTimeout(r, 50)", () => {
+    const content = "await new Promise((r) => setTimeout(r, 50));";
+    const vs = findViolations(content);
+    expect(vs).toHaveLength(1);
```
[the remark this answers is not in the document]
Comment (theshadow27) [reply scripts/check-test-timeouts.spec.ts:65]: The PR added 6 dedicated findViolations tests in the new describe block (lines 51–101 of the spec). The existing 19-case matrix is correctly scoped to hasFixedDelay, which remains exported and tested.

File: scripts/check-test-timeouts.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (theshadow27) [reply scripts/check-test-timeouts.ts]: Fixed in 0b16d1f: extracted `extractDelayArg` which locates the delay via first/second top-level comma (not the last comma), so `setTimeout(fn, 50, arg)` is correctly flagged (delay=50) and `setTimeout(fn, TIMEOUT, 50)` is correctly allowed (delay=TIMEOUT). Both `hasFixedDelay` and `findViolations` share the helper. Two new 3-arg test cases added to the matrix.

File: scripts/check-test-timeouts.spec.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (theshadow27) [reply scripts/check-test-timeouts.spec.ts:65]: Fixed in 0b16d1f: moved `shouldMatch`/`shouldNotMatch` to module scope and added a parallel `findViolations (single-line equivalence)` describe block that runs the full 21-case matrix through `findViolations`. The production scan path (which calls `findViolations`, not `hasFixedDelay`) is now directly regression-tested for comments, spacing, word-boundary guards, named constants, and arity edge cases.

**Comment by theshadow27** [comment]
## QA Verification

**PR:** #1985 — fix(ci): catch multi-line and callback setTimeout patterns in check-test-timeouts (fixes #1632, fixes #1633)
**Linked issues:** #1632, #1633

### Implementation Evidence

- `scripts/check-test-timeouts.ts` — introduces `extractDelayArg(args)` (lines 44–62) that extracts the 2nd positional argument as the delay, correctly handling 3-arg `setTimeout(fn, 50, extraArg)` forms (Copilot thread addressed in commit `0b16d1f`). `hasFixedDelay` and `findViolations` both use this helper.
- `scripts/check-test-timeouts.ts:findViolations` (lines 98–130) — new exported function that scans full file content with parenthesis-depth tracking across newlines, catching multi-line `setTimeout` calls (fixes #1633).
- `scripts/check-test-timeouts.ts:scanDir` — delegates to `findViolations` instead of the old line-by-line split loop.
- `scripts/check-test-timeouts.ts:hasFixedDelay` — retained and updated to use `extractDelayArg`; existing per-line unit tests continue to pass.
- `packages/daemon/src/claude-session/ws-server.spec.ts:1016` and `:3305` — confirmed replaced with `pollUntil` pattern; no fixed-delay `setTimeout` violations remain.

### Test Coverage

- `scripts/check-test-timeouts.spec.ts` — 48 tests (3 describe blocks):
  - `hasFixedDelay` matrix: 11 `shouldMatch` + 10 `shouldNotMatch` cases including arrow-fn callbacks, 3-arg form, and numeric literal forms
  - `findViolations (single-line equivalence)`: same 21-case matrix run through `findViolations`, confirming the production scan path
  - `findViolations (multi-line)`: 6 cases covering multi-line delay argument, arrow-fn callback on next line, variable delay (no flag), single-arg (no flag), correct line numbers for multiple violations
- `bun scripts/check-test-timeouts.ts` — no violations found in codebase

### Test Results

```
25 → 48 tests pass (0 fail) — scripts/check-test-timeouts.spec.ts
6835 pass, 0 fail — full suite (287 files)
```

### CI Status

All 3 checks **pass** on run [25298295524](https://github.com/theshadow27/mcp-cli/actions/runs/25298295524):
- `check` — pass (4m2s)
- `coverage` — pass (6m15s)
- `build` — pass (11s)

### Copilot Threads

2 threads dismissed inline:
- Thread on `check-test-timeouts.ts:121` (last-comma miss for 3-arg form) — addressed by commit `0b16d1f` which switched to `extractDelayArg` (2nd-arg extraction)
- Thread on `check-test-timeouts.spec.ts:52` (coverage gap for `findViolations`) — addressed by the new single-line equivalence describe block in commit `0b16d1f`

### Remaining Work (non-blocking)

None identified. Both #1632 and #1633 scopes are fully addressed.

### Outcome

**qa:pass** — All described functionality implemented and verified. Tests and CI green.
